tablo/geom_utils.py

```diff
@@ -437,4 +437,196 @@ def extract_significant_digits(number):
+
+            intersects = False
+            for r in reversed(outer_rings):     # TODO why esri loops over a reversed rings?
+                outer_ring = r[0]
+                if Terraformer.arrays_intersect_arrays(outer_ring, hole):
+                    outer_rings[r].push(hole)
+                    intersects = True
+                    break
+
+            # hole does not intersect ANY outer ring at this point; make it an outer ring.
+            if not intersects:
+                hole.reverse()
+                outer_rings.append([hole])
+
+        return {
+            'type': 'MultiPolygon',
+            'coordinates': outer_rings
+        }
+
+    @staticmethod
+    def close_ring(coordinates):
+        # makes sure the ring is closed
+        if coordinates[0] != coordinates[-1]:
+            coordinates.push(coordinates[0])
+
+        return coordinates
+
+    @staticmethod
+    def ring_is_clockwise(ring_to_test):
+        # a positive sum of areas under each side of a polygon indicates a clockwise polygon
+        # http://stackoverflow.com/questions/1165647/how-to-determine-if-a-list-of-polygon-points-are-in-clockwise-order
+        total = 0
+        r_length = len(ring_to_test)
+        for idx, pt1 in enumerate(ring_to_test):
+            if idx + 1 != r_length:
+                pt2 = ring_to_test[idx + 1]
+                total += (pt2[0] - pt1[0]) * (pt2[1] + pt1[1])
+        return total >= 0
+
+    @staticmethod
+    def edge_intersects_edge(a1, a2, b1, b2):
+        ua_t = (b2[0] - b1[0]) * (a1[1] - b1[1]) - (b2[1] - b1[1]) * (a1[0] - b1[0])
+        ub_t = (a2[0] - a1[0]) * (a1[1] - b1[1]) - (a2[1] - a1[1]) * (a1[0] - b1[0])
+        u_b = (b2[1] - b1[1]) * (a2[0] - a1[0]) - (b2[0] - b1[0]) * (a2[1] - a1[1])
+
+        if u_b != 0:
+            ua = ua_t / u_b
+            ub = ub_t / u_b
+
+            if 0 <= ua <= 1 and 0 <= ub <= 1:
+                return True
+
+        return False
+
+    @staticmethod
+    def arrays_intersect_arrays(a, b):
+        try:
+            # test for numbers
+            int(a[0][0])
+            try:
+                int(b[0][0])
+                for i, _ in enumerate(a[:-1]):
+                    for j, __ in enumerate(b[:-1]):
+                        if Terraformer.edge_intersects_edge(a[i], a[i + 1], b[j], b[j + 1]):
+                            return True
+            except TypeError:
+                for k in b:
+                    if Terraformer.arrays_intersect_arrays(a, k):
+                        return True
+        except TypeError:
+            for l in a:
+                if Terraformer.arrays_intersect_arrays(l, b):
+                    return True
+        return False
+
+    @staticmethod
+    def coordinates_contain_point(coordinates, point):
+        contains = False
+        i = 0
+        l = len(coordinates)
+        j = l - 1
+        while i < l:
+            if (((coordinates[i][1] <= point[1] < coordinates[j][1]) or (
```

Comment:
It would probably be a good idea to break out the components of this condition into separate variables within the loop. It is difficult to understand both what is being compared and the logic of the comparison.